### ISSUE TITLE
Unable to perform predict_proba when trained SetFit Body + Head

### ISSUE THREAD

**Issue by Swty13** [huggingface/setfit#261]
Hi Team,

I have trained SetFit trainer for text classification task on my custom data as following :
`model = SetFitModel.from_pretrained("sentence-transformers/paraphrase-mpnet-base-v2",
    use_differentiable_head=True,
    head_params={"out_features": num_classes},
)
trainer = SetFitTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=valid_dataset,
    loss_class=CosineSimilarityLoss,
    metric="accuracy",
    batch_size=16,
    num_iterations=20, # The number of text pairs to generate for contrastive learning
    num_epochs=1, # The number of epochs to use for contrastive learning
    column_mapping={"text": "text", "label": "label"} # Map dataset columns to text/label expected by trainer
)

trainer.unfreeze(keep_body_frozen=False)
trainer.train(
    num_epochs=25, # The number of epochs to train the head or the whole model (body and head)
    batch_size=8,
    body_learning_rate=1e-5, # The body's learning rate
    learning_rate=1e-2 ,# The head's learning rate
    l2_weight=0.00, # Weight decay on **both** the body and head. If `None`, will use 0.01.
)
metrics = trainer.evaluate()`

To get predicted label
`predict=model.predict(test_data['review'].tolist())`

**To get predicted probabilities : Below line didn't execute and throws error**
`proba=model.predict_proba(test_data['review'].tolist())`

**Error**
`TypeError: linear(): argument 'input' (position 1) must be Tensor, not numpy.ndarray`

**Model configuration** 
`{'model_body': SentenceTransformer(
   (0): Transformer({'max_seq_length': 512, 'do_lower_case': False}) with Transformer model: MPNetModel 
   (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
 ),
 'model_head': SetFitHead({'in_features': 768, 'out_features': 2, 'temperature': 1.0, 'bias': True, 'device': 'cuda'}),
 'multi_target_strategy': None,
 'l2_weight': 0.01,
 'normalize_embeddings': False}
`

#########################################################################################
#########################################################################################
Also , when I trained again as following it works fine for me

`
model1 = SetFitModel.from_pretrained("sentence-transformers/paraphrase-mpnet-base-v2")
trainer = SetFitTrainer(
    model=model1,
    train_dataset=train_dataset,
    eval_dataset=valid_dataset,
    loss_class=CosineSimilarityLoss,
    metric="accuracy",
    batch_size=8,
    num_iterations=20, # The number of text pairs to generate for contrastive learning
    num_epochs=3, # The number of epochs to use for constrastive learning
    column_mapping={"text": "text", "label": "label"} # Map dataset columns to text/label expected by trainer
)
trainer.train()
metrics = trainer.evaluate()
`

It will give both predicted labels and probabilities

`predict=model.predict(test_data['review'].tolist())
proba=model.predict_proba(test_data['review'].tolist())
`
**Model Configuration**

`{'model_body': SentenceTransformer(
   (0): Transformer({'max_seq_length': 512, 'do_lower_case': False}) with Transformer model: MPNetModel 
   (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
 ),
 'model_head': LogisticRegression(),
 'multi_target_strategy': None,
 'l2_weight': 0.01,
 'normalize_embeddings': False}
`

It seems when I train body + head of model I am unable to get probabilities of predicted labels. Please suggest what I was doing wrong here.
Thanks in advance !

**Comment by tomaarsen** [comment]
Hello!

Currently, `predict_proba` is failing when the differentiable head is used. See the following lines for the culprit:
https://github.com/huggingface/setfit/blob/fe0e7f4e1118cd8788fccdf33973dfa1d601bd6d/src/setfit/modeling.py#L381-L383
Note that `self.model_body.encode` returns an np.ndarray, and that `self.model_head.predict_proba` expects a torch Tensor when the head is the differentiable `SetFitHead`. 
#207 exists to tackle this issue, but it introduces some issues of its own.

I suggest to use the following workaround:
```python
model = ...

def predict_proba(model: SetFitModel):
    embeddings = model.model_body.encode(x_test, normalize_embeddings=self.normalize_embeddings, convert_to_tensor=True)
    return model.model_head.predict_proba(embeddings)

probs = predict_proba(model)
```

I apologize for the inconvenience. 

- Tom Aarsen

**Comment by tomaarsen** [comment]
Solved via #207. Thanks for reporting!

**Comment by Swty13** [comment]
@tomaarsen Thanks this works .
